src/protocols/vmps/vmps.c

```diff
@@ -262,60 +263,59 @@ static int contents[5][VQP_MAX_ATTRIBUTES] = {
 };
 #endif
 
-ssize_t fr_vmps_encode(uint8_t *buffer, size_t buflen, uint8_t const *original,
+ssize_t fr_vmps_encode(fr_dbuff_t *dbuff, uint8_t const *original,
 		       int code, uint32_t seq_no, fr_cursor_t *cursor)
 {
-	uint8_t *attr;
-	fr_pair_t *vp;
+	fr_dbuff_t		work_dbuff = FR_DBUFF_NO_ADVANCE(dbuff);
+	fr_pair_t 		*vp;
+	fr_dbuff_marker_t	hdr;
+	uint32_t 		sequence;
 
-	if (buflen < 8) {
-		fr_strerror_printf("Output buffer is too small for VMPS header. (%zu < 8)", buflen);
-		return -1;
-	}
+	/*
+	 *	Verify space for the packet...
+	 */
+	FR_DBUFF_CHECK_REMAINING_RETURN(&work_dbuff, FR_VQP_HDR_LEN);
+
+	/*
+	 *	Let's keep reference for packet header.
+	 */
+	fr_dbuff_marker(&hdr, &work_dbuff);
 
 	/*
 	 *	Create the header
 	 */
-	buffer[0] = FR_VQP_VERSION;	/* Version */
-	buffer[1] = code;		/* Opcode */
-	buffer[2] = FR_ERROR_CODE_VALUE_NO_ERROR;	/* Response Code */
-	buffer[3] = 0;			/* Data Count */
+	fr_dbuff_bytes_in(&work_dbuff, FR_VQP_VERSION,			/* Version */
+					code,				/* Opcode */
+					FR_ERROR_CODE_VALUE_NO_ERROR,	/* Response Code */
+					0);				/* Data Count */
 
 	if (original) {
-		memcpy(buffer + 4, original + 4, 4);
+		fr_dbuff_memcpy_in(&work_dbuff, original + 4, 4);
 	} else {
-		seq_no = htonl(seq_no);
-		memcpy(buffer + 4, &seq_no, 4);
+		fr_dbuff_in(&work_dbuff, seq_no);
 	}
 
-	attr = buffer + 8;
-
 	/*
 	 *	Encode the VP's.
 	 */
 	while ((vp = fr_cursor_current(cursor))) {
 		size_t len;
 
-		if (attr >= (buffer + buflen)) break;
-
 		if (vp->da == attr_packet_type) {
-			buffer[1] = vp->vp_uint32;
+			fr_dbuff_marker_current(&hdr)[1] = (uint8_t)vp->vp_uint32;
```

Comment:
Yeah this isn't great... but you don't have any option because the 'in' functions don't work with markers yet.  Hopefully this can be fixed in future.